README.md
spocc
========



Linux: [![Build Status](https://api.travis-ci.org/ropensci/spocc.png)](https://travis-ci.org/ropensci/spocc)  
Windows: [![Build status](https://ci.appveyor.com/api/projects/status/3d43armi2oanva2s)](https://ci.appveyor.com/project/karthik/spocc)  

**`spocc` = SPecies OCCurrence data**

At rOpenSci, we have been writing R packages to interact with many sources of species occurrence data, including [GBIF][gbif], [Vertnet][vertnet], [BISON][bison], [iNaturalist][inat], the [Berkeley ecoengine][ecoengine], and [AntWeb][antweb]. `spocc` is an R package to query and collect species occurrence data from many sources. The goal is to wrap functions in other R packages to make a seamless experience across data sources for the user.

The inspiration for this comes from users requesting a more seamless experience across data sources, and from our work on a similar package for taxonomy data ([taxize][taxize]).

__BEWARE:__ In cases where you request data from multiple providers, especially when including GBIF, there could be duplicate records since many providers' data eventually ends up with GBIF. See `?spocc_duplicates`, after installation, for more.

## Contributing

See [CONTRIBUTING.md](CONTRIBUTING.md)

## Installation

Install `spocc`


```r
install.packages("spocc", dependencies = TRUE)
```

Or the development version


```r
install.packages("devtools")
devtools::install_github("ropensci/spocc")
```


```r
library("spocc")
```

## Get data

Get data from GBIF


```r
(out <- occ(query='Accipiter striatus', from='gbif', limit=100))
#> Summary of results - occurrences found for: 
#>  gbif  : 100 records across 1 species 
#>  bison :  0 records across 1 species 
#>  inat  :  0 records across 1 species 
#>  ebird :  0 records across 1 species 
#>  ecoengine :  0 records across 1 species 
#>  antweb :  0 records across 1 species
```


```r
out$gbif # just gbif data
#> Species [Accipiter striatus (100)] 
#> First 10 rows of [Accipiter_striatus]
#> 
#>                  name  longitude latitude prov                 issues
#> 1  Accipiter striatus  -71.72514 18.26982 gbif    cdround,cudc,gass84
#> 2  Accipiter striatus  -72.52547 43.13234 gbif    cdround,cudc,gass84
#> 3  Accipiter striatus  -97.00035 33.07049 gbif    cdround,cudc,gass84
#> 4  Accipiter striatus  -97.19930 32.86027 gbif    cdround,cudc,gass84
#> 5  Accipiter striatus  -97.65347 30.15791 gbif    cdround,cudc,gass84
#> 6  Accipiter striatus -122.43980 37.48967 gbif    cdround,cudc,gass84
#> 7  Accipiter striatus  -76.37695 42.42883 gbif                cdround
#> 8  Accipiter striatus  -76.33708 42.25353 gbif                cdround
#> 9  Accipiter striatus  -99.84577 20.62069 gbif    cdround,cudc,gass84
#> 10 Accipiter striatus -119.25619 34.23091 gbif cdround,gass84,mdatunl
#> ..                ...        ...      ...  ...                    ...
#> Variables not shown: key (int), datasetKey (chr), publishingOrgKey (chr),
#>      publishingCountry (chr), protocol (chr), lastCrawled (chr),
#>      lastParsed (chr), extensions (chr), basisOfRecord (chr), taxonKey
#>      (int), kingdomKey (int), phylumKey (int), classKey (int), orderKey
#>      (int), familyKey (int), genusKey (int), speciesKey (int),
#>      scientificName (chr), kingdom (chr), phylum (chr), order (chr),
#>      family (chr), genus (chr), species (chr), genericName (chr),
#>      specificEpithet (chr), taxonRank (chr), dateIdentified (chr), year
#>      (int), month (int), day (int), eventDate (chr), modified (chr),
#>      lastInterpreted (chr), references (chr), identifiers (chr), facts
#>      (chr), relations (chr), geodeticDatum (chr), class (chr), countryCode
#>      (chr), country (chr), verbatimEventDate (chr), verbatimLocality
#>      (chr), rights (chr), rightsHolder (chr), occurrenceID (chr),
#>      collectionCode (chr), taxonID (chr), gbifID (chr), occurrenceRemarks
#>      (chr), institutionCode (chr), catalogNumber (chr), datasetName (chr),
#>      recordedBy (chr), occurrenceDetails (chr), identifier (chr),
#>      identificationID (chr), informationWithheld (chr), eventTime (chr),
#>      sex (chr), continent (chr), stateProvince (chr), georeferencedDate
#>      (chr), institutionID (chr), higherGeography (chr), type (chr),
#>      identifiedBy (chr), georeferenceSources (chr),
#>      identificationVerificationStatus (chr), samplingProtocol (chr),
#>      endDayOfYear (chr), otherCatalogNumbers (chr), preparations (chr),
#>      georeferenceVerificationStatus (chr), nomenclaturalCode (chr),
#>      individualID (chr), higherClassification (chr), locationAccordingTo
#>      (chr), previousIdentifications (chr), verbatimCoordinateSystem (chr),
#>      georeferenceProtocol (chr), identificationQualifier (chr),
#>      accessRights (chr), dynamicProperties (chr), county (chr), locality
#>      (chr), language (chr), georeferencedBy (chr), lifeStage (chr),
#>      establishmentMeans (chr), startDayOfYear (chr), occurrenceStatus
#>      (chr), elevation (dbl), elevationAccuracy (dbl), waterBody (chr),
#>      recordNumber (chr), samplingEffort (chr), locationRemarks (chr),
#>      vernacularName (chr)
```

Get fine-grained detail over each data source by passing on parameters to the packge rebird in this example.


```r
out <- occ(query='Setophaga caerulescens', from='ebird', ebirdopts=list(region='US'))
out$ebird # just ebird data
#> Species [Setophaga caerulescens (16)] 
#> First 10 rows of [Setophaga_caerulescens]
#> 
#>                      name longitude latitude  prov
#> 1  Setophaga caerulescens -80.82018 24.81497 ebird
#> 2  Setophaga caerulescens -80.31086 25.73408 ebird
#> 3  Setophaga caerulescens -80.27264 25.68065 ebird
#> 4  Setophaga caerulescens -80.92550 25.14160 ebird
#> 5  Setophaga caerulescens -80.12498 25.81765 ebird
#> 6  Setophaga caerulescens -80.30560 25.62582 ebird
#> 7  Setophaga caerulescens -80.34860 25.55849 ebird
#> 8  Setophaga caerulescens -80.12554 26.38334 ebird
#> 9  Setophaga caerulescens -80.27710 26.06974 ebird
#> 10 Setophaga caerulescens -80.15430 25.94190 ebird
#> ..                    ...       ...      ...   ...
#> Variables not shown: comName (chr), howMany (int), locID (chr), locName
#>      (chr), locationPrivate (lgl), obsDt (chr), obsReviewed (lgl),
#>      obsValid (lgl)
```

Get data from many sources in a single call


```r
ebirdopts = list(region='US'); gbifopts = list(country='US')
out <- occ(query='Setophaga caerulescens', from=c('gbif','bison','inat','ebird'), gbifopts=gbifopts, ebirdopts=ebirdopts, limit=50)
head(occ2df(out)); tail(occ2df(out))
#>                     name longitude latitude prov
#> 1 Setophaga caerulescens -72.46131 44.34088 gbif
#> 2 Setophaga caerulescens -72.96430 44.20737 gbif
#> 3 Setophaga caerulescens -73.11068 44.32922 gbif
#> 4 Setophaga caerulescens -72.93318 43.44171 gbif
#> 5 Setophaga caerulescens -70.92961 42.41968 gbif
#> 6 Setophaga caerulescens -71.14434 42.37231 gbif
#>                       name longitude latitude  prov
#> 161 Setophaga caerulescens -80.16017 26.48675 ebird
#> 162 Setophaga caerulescens -80.32015 25.60671 ebird
#> 163 Setophaga caerulescens -80.47223 25.06672 ebird
#> 164 Setophaga caerulescens -80.16164 25.90072 ebird
#> 165 Setophaga caerulescens -80.16423 25.90170 ebird
#> 166 Setophaga caerulescens -80.37937 25.75437 ebird
```

## Make maps

### Leaflet


```r
spp <- c('Danaus plexippus','Accipiter striatus','Pinus contorta')
dat <- occ(query = spp, from = 'gbif', gbifopts = list(hasCoordinate=TRUE))
data <- occ2df(dat)
mapleaflet(data = data, dest = ".")
```

![leafletmap](http://f.cl.ly/items/3w2Y1E3Z0T2T2z40310K/Screen%20Shot%202014-02-09%20at%2010.38.10%20PM.png)


### Github gist


```r
spp <- c('Danaus plexippus','Accipiter striatus','Pinus contorta')
dat <- occ(query=spp, from='gbif', gbifopts=list(hasCoordinate=TRUE))
dat <- fixnames(dat)
dat <- occ2df(dat)
mapgist(data=dat, color=c("#976AAE","#6B944D","#BD5945"))
```

![gistmap](http://f.cl.ly/items/343l2G0A2J3T0n2t433W/Screen%20Shot%202014-02-09%20at%2010.40.57%20PM.png)


### ggplot2


```r
ecoengine_data <- occ(query = 'Lynx rufus californicus', from = 'ecoengine')
mapggplot(ecoengine_data)
```

![ggplot2map](http://f.cl.ly/items/1U1R0E0G392l2q362V33/Screen%20Shot%202014-02-09%20at%2010.44.59%20PM.png)


### Base R plots


```r
spnames <- c('Accipiter striatus', 'Setophaga caerulescens', 'Spinus tristis')
out <- occ(query=spnames, from='gbif', gbifopts=list(hasCoordinate=TRUE))
plot(out, cex=1, pch=10)
```

![basremap](http://f.cl.ly/items/3O13330W3w3Z0H3u1X0s/Screen%20Shot%202014-02-09%20at%2010.46.25%20PM.png)

## Meta

* Please report any issues or bugs](https://github.com/ropensci/spocc/issues).
* License: MIT
* Get citation information for `spocc` in R doing `citation(package = 'spocc')`

[![ropensci footer](http://ropensci.org/public_images/github_footer.png)](http://ropensci.org)

[gbif]: https://github.com/ropensci/rgbif
[vertnet]: https://github.com/ropensci/rvertnet
[bison]: https://github.com/ropensci/rbison
[inat]: https://github.com/ropensci/rinat
[taxize]: https://github.com/ropensci/taxize
[ecoengine]: https://github.com/ropensci/ecoengine
[antweb]: http://antweb.org/


NEWS
spocc 0.2.4
===============

MINOR IMPROVEMENTS

* Improved documentation for bounding boxes, their expected format, etc. (#96)
* Remove dependency on the following packages: `assertthat`, `plyr`, `data.table`, and `XML` (#102)
* Using package `gistr` now to post interactive geojson maps on GitHub gists (#100)
* `rgbif` now must be `v0.7.7` or greater (the latest version on CRAN).
* Removed the startup message.

BUG FIXES

* Duplicate, but not working correctly, function `occ2sp()` removed. The function `occ_to_sp()` function is the working version. (#97)
* Fixed bug where some records returned form GBIF did not have lat/long column headers, and we internally rearranged columns, which caused complete stop when that happened. Fixed now. (#101)
* Changed all `\donttest` to `\dontrun` in examples as requested by CRAN maintainers (#99)

spocc 0.2.2
===============

NEW FEATURES

* Added new function `occ_names()` to search only for taxonomic names. The goal here is to use ths function if there is some question about what names you want to use to search for occurrences with. (#84). Suggested by @jarioksa
* New function `occ_names_options()` to quickly get parameter options to pass to `occ_names()`.
* New `summary()` method for the `occdat` `S3` object that is output from `occ()` (#83)
* In many places in `spocc` (README, vignette, `occ()` documentation file, at package startup), we make it clear that there could be duplicate records returned in certain scenarios. And a new documentation page detailing what to watch out for: `?spocc_duplicates`. (#77)

MINOR IMPROVEMENTS

* All latitude/longitude column headers are now changed to latitude and longitude, whereas they use to vary from `latitude`, `decimalLatitude`, `Latitude`, `lat`, and `decimal_latitude`. (#91)
* Default is 500 now for the `limit` parameter in `occ()` (#78)
* You can now pass in `limit` to each functions options parameter, and it will work. Each data source can have a different parameter internally from `limit`, but now internally within `spocc`, we allow you to use `limit` so you don't have to know what the data source specific parameter is. (#81)
* There is a now a startup message to give information on the package (#79)
* `occ_options()` gains new parameter `where` to print either in the console or to open man file in the IDE, or prints to console in command line R. 

spocc 0.2.0
===============

NEW FEATURES

* `occ()` gains new parameter `callopts` to pass on curl debugging options to `httr::GET()` (#35)
* `wkt_vis()` now by default plots a well known text area (WKT) on an interactive mapbox map in your default browser. New parameter `which` allows you to choose the interactive map or a static ggplot2 map. (#70)
* Individual data sources `occ()` gains new class. In the previous version of this package, a `data.frame` was printed. Now the data is assigned the object `occdatind` (short for _occdat individual_).
* `occ()` now uses a print method for the `occdatind` class, adopted from `dplyr` that prints a brief `data.frame`, with columns wrapped to fit the width of your console, and additional columns not printed given at bottom with their class type. Note that the print behavior for the resulting object of an `occ()` call remains the same. (#69) (#74)

MINOR IMPROVEMENTS

* Added `whisker` as a package import to use in the `wkt_vis()` function. (#70)
* Mapping functions now all accept the same input. Previously `mapggplot()` accepted the output of `occ()`, of class `occdat`, while the other two functions for mapping, `mapleaflet()` and `mapgist()` accepted a `data.frame`. Now all three functions accept  the output of `occ()`, an object of class `occdat`. (#75)
* The `meta` slot in each returned object (indexed by `object$meta`) contains spots for `returned` and `found`, to designate number of records returned, and number of records found. (#64)

BUG FIXES

* Fixed bug in AntWeb output, where there was supposed to be a column titled `name`. (#71)

spocc 0.1.4
===============

NEW FEATURES

* Can now do geometry only queries. See examples in `occ()`.
* In addition, you can pass in sp objects of SpatialPolygons or SpatialPolygonsDataFrame classes.

spocc 0.1.2
===============

NEW FEATURES

* There were quite a few changes in one of the key packages that `spocc` depends on: `rgbif`. A number of input and output parameter names changed. A new version of `rgbif` was pushed to CRAN. (#56)
* New function `clean_spocc()` started (not finished yet) to attempt to clean data. For example, one use case is removing impossible lat/long values (i.e., longitue values greater than absolute 180). Another, not implemented yet, is to remove points that are not in the country or habitat your points are supposed to be in. (#44)
* New function `fixnames()` to trim species names with optional input parameters to make data easier to use for mapping.
* New function `wkt_vis()` to visualize a WKT (well-known text) area on a map. Uses `ggmap` to pull down a Google map so that the visualization has some geographic and natural earth context. We'll soon introduce an interactive version of this function that will bring up a small Shiny app to draw a WKT area, then return those coordinates to your R session. (#34)

MINOR IMPROVEMENTS

* Added a CONTRIBUTING.md file to the github repo to help guide contributions (#61)
* Packages that require a certain version are forced to be X version or greater. Thes are rinat (>= 0.1.1), rbison (>= 0.3.2), rgbif (>= 0.6.2), ecoengine (>= 1.3), rebird (>= 0.1.1), AntWeb (>= 0.6.1), and leafletR (>= 0.2-0). This should help avoid problems.
* General improvement to function documentation.

spocc 0.1.0
===============

* Initial release to CRAN


R/occ.r
#' Search for species occurrence data across many data sources.
#'
#' Search on a single species name, or many. And search across a single
#' or many data sources.
#'
#' @import rgbif rinat rebird ecoengine rbison AntWeb
#' @importFrom lubridate now
#' @template occtemp
#' @export
#' @examples \dontrun{
#' # Single data sources
#' (res <- occ(query = 'Accipiter striatus', from = 'gbif', limit = 50))
#' res$gbif
#' (res <- occ(query = 'Accipiter striatus', from = 'ecoengine', limit = 50))
#' res$ecoengine
#' (res <- occ(query = 'Accipiter striatus', from = 'ebird', limit = 50))
#' res$ebird
#' (res <- occ(query = 'Danaus plexippus', from = 'inat', limit = 50))
#' res$inat
#' (res <- occ(query = 'Bison bison', from = 'bison', limit = 50))
#' res$bison
#' # Data from AntWeb
#' # By species
#' (by_species <- occ(query = "linepithema humile", from = "antweb"))
#' # or by genus
#' (by_genus <- occ(query = "acanthognathus", from = "antweb"))
#'
#' occ(query = 'Setophaga caerulescens', from = 'ebird', ebirdopts = list(region='US'))
#' occ(query = 'Spinus tristis', from = 'ebird', ebirdopts =
#'    list(method = 'ebirdgeo', lat = 42, lng = -76, dist = 50))
#'
#' # You can pass on limit param to all sources even though its a different param in that source
#' ## ecoengine example
#' res <- occ(query = 'Accipiter striatus', from = 'ecoengine', ecoengineopts=list(limit = 5))
#' res$ecoengine
#' ## This is particularly useful when you want to set different limit for each source
#' res <- occ(query = 'Accipiter striatus', from = c('gbif','ecoengine'),
#'    gbifopts=list(limit = 10), ecoengineopts=list(limit = 5))
#' res
#'
#' # Many data sources
#' (out <- occ(query = 'Pinus contorta', from=c('gbif','bison')))
#'
#' ## Select individual elements
#' out$gbif
#' out$gbif$data
#'
#' ## Coerce to combined data.frame, selects minimal set of columns (name, lat, long)
#' occ2df(out)
#'
#' # Pass in limit parameter to all sources. This limits the number of occurrences
#' # returned to 10, in this example, for all sources, in this case gbif and inat.
#' occ(query='Pinus contorta', from=c('gbif','inat'), limit=10)
#'
#' # Geometry
#' ## Pass in geometry parameter to all sources. This constraints the search to the
#' ## specified polygon for all sources, gbif and bison in this example.
#' ## Check out http://arthur-e.github.io/Wicket/sandbox-gmaps3.html to get a WKT string
#' occ(query='Accipiter', from='gbif',
#'    geometry='POLYGON((30.1 10.1, 10 20, 20 60, 60 60, 30.1 10.1))')
#' occ(query='Helianthus annuus', from='bison',
#'    geometry='POLYGON((-111.06 38.84, -110.80 39.37, -110.20 39.17, -110.20 38.90,
#'                       -110.63 38.67, -111.06 38.84))')
#'
#' ## Or pass in a bounding box, which is automatically converted to WKT (required by GBIF)
#' ## via the bbox2wkt function. The format of a bounding box is
#' ## [min-longitude, min-latitude, max-longitude, max-latitude].
#' occ(query='Accipiter striatus', from='gbif', geometry=c(-125.0,38.4,-121.8,40.9))
#'
#' ## Bounding box constraint with ecoengine
#' ## Use this website: http://boundingbox.klokantech.com/ to quickly grab a bbox.
#' ## Just set the format on the bottom left to CSV.
#' occ(query='Accipiter striatus', from='ecoengine', limit=10,
#'    geometry=c(-125.0,38.4,-121.8,40.9))
#'
#' ## lots of results, can see how many by indexing to meta
#' res <- occ(query='Accipiter striatus', from='gbif',
#'    geometry='POLYGON((-69.9 49.2,-69.9 29.0,-123.3 29.0,-123.3 49.2,-69.9 49.2))')
#' res$gbif
#'
#' ## You can pass in geometry to each source separately via their opts parameter, at
#' ## least those that support it. Note that if you use rinat, you reverse the order, with
#' ## latitude first, and longitude second, but here it's the reverse for consistency across
#' ## the spocc package
#' bounds <- c(-125.0,38.4,-121.8,40.9)
#' occ(query = 'Danaus plexippus', from="inat", geometry=bounds)
#'
#' ## Passing geometry with multiple sources
#' occ(query = 'Danaus plexippus', from=c("inat","gbif","ecoengine"), geometry=bounds)
#'
#' ## Using geometry only for the query
#' ### A single bounding box
#' occ(geometry = bounds, from = "gbif")
#' ### Many bounding boxes
#' occ(geometry = list(c(-125.0,38.4,-121.8,40.9), c(-115.0,22.4,-111.8,30.9)), from = "gbif")
#'
#' # Specify many data sources, another example
#' ebirdopts = list(region = 'US'); gbifopts  =  list(country = 'US')
#' out <- occ(query = 'Setophaga caerulescens', from = c('gbif','inat','bison','ebird'),
#'     gbifopts = gbifopts, ebirdopts = ebirdopts)
#' occ2df(out)
#'
#' # Pass in many species names, combine just data to a single data.frame, and
#' # first six rows
#' spnames <- c('Accipiter striatus', 'Setophaga caerulescens', 'Spinus tristis')
#' out <- occ(query = spnames, from = 'gbif', gbifopts = list(hasCoordinate = TRUE))
#' df <- occ2df(out)
#' head(df)
#'
#' # taxize integration
#' ## You can pass in taxonomic identifiers
#' library("taxize")
#' (ids <- get_ids(names=c("Chironomus riparius","Pinus contorta"), db = c('itis','gbif')))
#' occ(ids = ids[[1]], from='bison')
#' occ(ids = ids, from=c('bison','gbif'), limit=20)
#'
#' (ids <- get_ids(names="Chironomus riparius", db = 'gbif'))
#' occ(ids = ids, from='gbif', limit=20)
#'
#' (ids <- get_gbifid("Chironomus riparius"))
#' occ(ids = ids, from='gbif', limit=20)
#'
#' (ids <- get_tsn('Accipiter striatus'))
#' occ(ids = ids, from='bison', limit=20)
#'
#' # SpatialPolygons/SpatialPolygonsDataFrame integration
#' library("sp")
#' ## Single polygon in SpatialPolygons class
#' one <- Polygon(cbind(c(91,90,90,91), c(30,30,32,30)))
#' spone = Polygons(list(one), "s1")
#' sppoly = SpatialPolygons(list(spone), as.integer(1))
#' out <- occ(geometry = sppoly, limit=50)
#' out$gbif$data
#'
#' ## Two polygons in SpatialPolygons class
#' one <- Polygon(cbind(c(-121.0,-117.9,-121.0,-121.0), c(39.4, 37.1, 35.1, 39.4)))
#' two <- Polygon(cbind(c(-123.0,-121.2,-122.3,-124.5,-123.5,-124.1,-123.0),
#'                      c(44.8,42.9,41.9,42.6,43.3,44.3,44.8)))
#' spone = Polygons(list(one), "s1")
#' sptwo = Polygons(list(two), "s2")
#' sppoly = SpatialPolygons(list(spone, sptwo), 1:2)
#' out <- occ(geometry = sppoly, limit=50)
#' out$gbif$data
#'
#' ## Two polygons in SpatialPolygonsDataFrame class
#' sppoly_df <- SpatialPolygonsDataFrame(sppoly, data.frame(a=c(1,2), b=c("a","b"), c=c(TRUE,FALSE),
#'    row.names=row.names(sppoly)))
#' out <- occ(geometry = sppoly_df, limit=50)
#' out$gbif$data
#'
#' # curl debugging
#' library('httr')
#' occ(query = 'Accipiter striatus', from = 'gbif', callopts=verbose())
#' occ(query = 'Accipiter striatus', from = 'ebird', callopts=verbose())
#' occ(query = 'Accipiter striatus', from = 'bison', callopts=verbose())
#' occ(query = 'Accipiter striatus', from = 'ecoengine', callopts=verbose())
#' occ(query = 'Accipiter striatus', from = c('ebird','bison'), callopts=verbose())
#' occ(query = 'Accipiter striatus', from = 'ebird', callopts=timeout(seconds = 0.1))
#' ## notice that callopts is ignored when from=inat or from=antweb
#' occ(query = 'Accipiter striatus', from = 'inat', callopts=verbose())
#' occ(query = 'linepithema humile', from = 'antweb', callopts=verbose())
#' }
#'
#' @examples \dontrun{
#' #### NOTE: no support for multipolygons yet
#' ## WKT's are more flexible than bounding box's. You can pass in a WKT with multiple
#' ## polygons like so (you can use POLYGON or MULTIPOLYGON) when specifying more than one
#' ## polygon. Note how each polygon is in it's own set of parentheses.
#' occ(query='Accipiter striatus', from='gbif',
#'    geometry='MULTIPOLYGON((30 10, 10 20, 20 60, 60 60, 30 10),
#'                           (30 10, 10 20, 20 60, 60 60, 30 10))')
#' }
occ <- function(query = NULL, from = "gbif", limit = 500, geometry = NULL, rank = "species",
    type = "sci", ids = NULL, callopts=list(), gbifopts = list(), bisonopts = list(),
    inatopts = list(), ebirdopts = list(), ecoengineopts = list(), antwebopts = list())
{
  if(!is.null(geometry)){
    if(class(geometry) %in% c('SpatialPolygons','SpatialPolygonsDataFrame')){
      geometry <- as.list(handle_sp(geometry))
    }
  }
  sources <- match.arg(from, choices = c("gbif", "bison", "inat", "ebird", "ecoengine", "antweb"),
                       several.ok = TRUE)
  if(!all(from %in% sources)){
    stop(sprintf("Woops, the following are not supported or spelled incorrectly: %s", from[!from %in% sources]))
  }

  loopfun <- function(x, y, z, w) {
    # x = query; y = limit; z = geometry; w = callopts
    gbif_res <- foo_gbif(sources, x, y, z, w, gbifopts)
    bison_res <- foo_bison(sources, x, y, z, w, bisonopts)
    inat_res <- foo_inat(sources, x, y, z, w, inatopts)
    ebird_res <- foo_ebird(sources, x, y, w, ebirdopts)
    ecoengine_res <- foo_ecoengine(sources, x, y, z, w, ecoengineopts)
    antweb_res <- foo_antweb(sources, x, y, z, w, antwebopts)
    list(gbif = gbif_res, bison = bison_res, inat = inat_res, ebird = ebird_res,
         ecoengine = ecoengine_res, antweb = antweb_res)
  }

  loopids <- function(x, y, z, w) {
    # x = query; y=limit; z=geometry
#     classes <- ifelse(length(x)>1, vapply(x, class, ""), class(x))
    classes <- class(x)
    if(!all(classes %in% c("gbifid","tsn")))
      stop("Currently, taxon identifiers have to be of class gbifid or tsn")
    if(class(x) == 'gbifid'){
      gbif_res <- foo_gbif(sources, x, y, z, w, gbifopts)
      bison_res <- list(time = NULL, data = data.frame(NULL))
    } else if(class(x) == 'tsn') {
      bison_res <- foo_bison(sources, x, y, z, w, bisonopts)
      gbif_res <- list(time = NULL, data = data.frame(NULL))
    }
    list(gbif = gbif_res,
         bison = bison_res,
         inat = list(time = NULL, data = data.frame(NULL)),
         ebird = list(time = NULL, data = data.frame(NULL)),
         ecoengine = list(time = NULL, data = data.frame(NULL)),
         antweb = list(time = NULL, data = data.frame(NULL)))
  }

  # check that one of query or ids is non-NULL
#   assert_that(xor(!is.null(query), !is.null(ids), !is.null(geometry)))
   if(!any(!is.null(query), !is.null(ids), !is.null(geometry)))
     stop("One of query, ids, or geometry parameters must be non-NULL")

  if(is.null(ids) && !is.null(query)){
    # If query not null (taxonomic names passed in)
    tmp <- lapply(query, loopfun, y=limit, z=geometry, w=callopts)
  } else if(is.null(query) && is.null(geometry)) {
    unlistids <- function(x){
      if(length(x) == 1){
        if(is.null(names(x))){ list(x) } else {
          if(!names(x) %in% c("gbif","itis"))
            list(x)
          else
            list(x[[1]])
        }
      } else {
        gg <- as.list(unlist(x, use.names = FALSE))
        hh <- as.vector(rep(vapply(x, class, ""), vapply(x, length, numeric(1))))
        if(all(hh == "character"))
          hh <- rep(class(x), length(x))
        for(i in seq_along(gg)){
          class(gg[[i]]) <- hh[[i]]
        }
        return( gg )
      }
    }
    ids <- unlistids(ids)
    # if ids is not null (taxon identifiers passed in)
    # ids can only be passed to gbif and bison for now
    # so don't pass anything on to ecoengine, inat, or ebird
    tmp <- lapply(ids, loopids, y=limit, z=geometry, w=callopts)
  } else {
    type <- 'geometry'
    if(is.numeric(geometry)){
      tmp <- list(loopfun(z=geometry, y=limit, x=query, w=callopts))
    } else if(is.list(geometry)){
      tmp <- lapply(geometry, function(b) loopfun(z=b, y=limit, x=query, w=callopts))
    }
  }

  getsplist <- function(srce, opts) {
    tt <- lapply(tmp, function(x) x[[srce]]$data)
    if(!is.null(query) && is.null(geometry)){ # query
      names(tt) <- gsub("\\s", "_", query)
      optstmp <- tmp[[1]][[srce]]$opts
    } else if(is.null(query) && !is.null(geometry)){ # geometry
#       if(is.numeric(geometry)){ gg <- paste(geometry,collapse=",") } else {
#         gg <- lapply(geometry, paste, collapse=",")
#       }
#       names(tt) <- gg
      tt <- tt
      optstmp <- tmp[[1]][[srce]]$opts
    } else if(!is.null(query) && !is.null(geometry)) { # query & geometry
      names(tt) <- gsub("\\s", "_", query)
      optstmp <- tmp[[1]][[srce]]$opts
    } else if(is.null(query) && is.null(geometry)) {
      names(tt) <- sapply(tmp, function(x) unclass(x[[srce]]$opts[[1]]))
      tt <- tt[!vapply(tt, nrow, 1) == 0]
      opts <- sc(lapply(tmp, function(x) x[[srce]]$opts))
      optstmp <- unlist(opts)
#       optstmp <- as.list(c(optstmp[!names(optstmp) %in% 'limit'], optstmp[names(optstmp) %in% 'limit'][1]))
#       optstmp <- as.list(c(optstmp[!names(optstmp) %in% 'count'], optstmp[names(optstmp) %in% 'count'][1]))
      simplist <- function(b){
        splitup <- unique(names(b))
        sapply(splitup, function(d){
          tmp <- b[names(b) %in% d]
          if(length(unique(unname(unlist(tmp)))) == 1){ as.list(tmp[1]) } else {
            outout <- list(unname(unlist(tmp)))
            names(outout) <- names(tmp)[1]
            outout
          }
        }, USE.NAMES=FALSE)
      }
      optstmp <- simplist(optstmp)
    }

    if (any(grepl(srce, sources))) {
      ggg <- list(meta = list(source = srce, time = tmp[[1]][[srce]]$time,
          found = tmp[[1]][[srce]]$found, returned = nrow(tmp[[1]][[srce]]$data),
          type = type, opts = optstmp), data = tt)
      class(ggg) <- "occdatind"
      ggg
    } else {
      ggg <- list(meta = list(source = srce, time = NULL, found = NULL, returned = NULL,
          type = NULL, opts = NULL), data = tt)
      class(ggg) <- "occdatind"
      ggg
    }
  }
  gbif_sp <- getsplist("gbif", gbifopts)
  bison_sp <- getsplist("bison", bisonopts)
  inat_sp <- getsplist("inat", inatopts)
  ebird_sp <- getsplist("ebird", ebirdopts)
  ecoengine_sp <- getsplist("ecoengine", ecoengineopts)
  antweb_sp <- getsplist("antweb", ecoengineopts)
  p <- list(gbif = gbif_sp, bison = bison_sp, inat = inat_sp, ebird = ebird_sp,
            ecoengine = ecoengine_sp, antweb = antweb_sp)
  class(p) <- "occdat"
  return(p)
}
